File: Blind150/5.Binary Search/Go/koko_eating_banan.go
// https://leetcode.com/problems/koko-eating-bananas/description/

// Helper function to determine if it's possible to eat all bananas in `h` hours at speed `k`
func canEat(piles []int, k, h int) bool {
	totalTime := 0
	for _, p := range piles {
		totalTime += int(math.Ceil(float64(p) / float64(k)))
	}
	return totalTime <= h
}

// Function to find the minimum eating speed that allows finishing all bananas within `h` hours
func minEatingSpeed(piles []int, h int) int {
	l, r := 1, maxInArray(piles)
	res := r

	for l <= r {
		mid := l + (r-l)/2
		if canEat(piles, mid, h) {
			res = min(res, mid)
			r = mid - 1
		} else {
			l = mid + 1
		}
	}

	return res
}


// Helper function to find the maximum value in a slice
func maxInArray(arr []int) int {
	m := arr[0]
	for _, v := range arr {
		if v > m {
			m = v
		}
	}
	return m
}

File: Blind150/5.Binary Search/Python/koko_eating_banan.py
# https://leetcode.com/problems/koko-eating-bananas/description/
from typing import List
from math import ceil

class Solution:
    def can_eat(self, piles, k, h):

        total_time = 0
        for p in piles:
            total_time += ceil(p/k)

        return total_time <= h


    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        l, r = 1, max(piles)

        res = r

        while l <= r:

            mid = l + (r-l) // 2
            can_eat = self.can_eat(piles, mid , h)

            if can_eat:
                res = min(res, mid)
                r = mid - 1
            else:
                l = mid + 1


        return res
